gufe/serialization/json.py

```diff
@@ -204,3 +209,147 @@ def serializer(self, obj: Any) -> str:
     def deserializer(self, string: str) -> Any:
         """Callable to loads JSON"""
         return self._deserializer(string)
+
+
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+# custom_codecs.py: A place to keep various custom JSONCodec instances
+
+import datetime
+import functools
+import pathlib
+from uuid import UUID
+
+import numpy as np
+from openff.units import DEFAULT_UNIT_REGISTRY
+
+import gufe
+from gufe.compression import zst_compress, zst_decompress
+from gufe.settings.models import SettingsBaseModel
+
+
+def default_from_dict(dct):
+    dct = dict(dct)  # make a copy
+    module = dct.pop("__module__")
+    qualname = dct.pop("__class__")
+    del dct[":is_custom:"]
+
+    cls = gufe.tokenization.get_class(module, qualname)
+    return cls(**dct)
+
+
+def inherited_is_my_dict(dct, cls):
+    if not ("__module__" in dct and "__class__" in dct):
+        return False
+    dct = dict(dct)
+    module = dct.pop("__module__")
+    classname = dct.pop("__class__")
+    stored = gufe.tokenization.get_class(module, classname)
+    return cls in stored.mro()
+
+
+def is_npy_dtype_dict(dct):
+    expected = ["dtype", "bytes"]
+    is_custom = all(exp in dct for exp in expected)
+    return is_custom and ("shape" not in dct)
+
+
+def is_openff_unit_dict(dct):
+    expected = ["pint_unit_registry", "unit_name", ":is_custom:"]
+    is_custom = all(exp in dct for exp in expected)
+    return is_custom and dct["pint_unit_registry"] == "openff_units"
+
+
+def is_openff_quantity_dict(dct):
+    expected = ["pint_unit_registry", "magnitude", ":is_custom:", "unit"]
+    is_custom = all(exp in dct for exp in expected)
+    return is_custom and dct["pint_unit_registry"] == "openff_units"
+
+
+PATH_CODEC = JSONCodec(
```

Comment:
If this is up for vote, I do generally like having the custom codes be in a separate file, both cleaner / more readable and makes it a lot less of a headache to find where things are.

Reply:
It definitely is up for a vote. Will update once we hear from others!

Reply:
I'm fine with the JSON components all being in one file. It gives parity with the `msgpack` components also all being in one place, and I'm not sure there's a whole lot of value in the separation we have now.

I often find myself entering the wrong file when I go looking for JSON things, then have to hop to the other one when I realize it.

Reply:
we'll keep it for now given the vote, but if it proves to be a headache we can always refactor.